### DISCUSSION TITLE
TBB and parallel stl?

### DISCUSSION THREAD

**Discussion by indigox4** [microsoft/STL#2512] [Q&A] [answered]
Can concurrent usage of TBB and the parallel versions of the stl algorithms in the same process cause thread oversubscription? Or is it possible they share the same thread pool?

For example, if a TBB parallel task calls a std::sort with a parallel execution policy, could this result in many more threads created than CPU cores available to run them?

**Comment by AlexGuteniev** [comment, marked as answer]
Regarding the number of chunks. There's a deliberate oversubscription
https://github.com/microsoft/STL/blob/bbd5dbaa61f192b393d12ea75a2dae1b8a014e7a/stl/inc/execution#L69-L71
https://github.com/microsoft/STL/blob/bbd5dbaa61f192b393d12ea75a2dae1b8a014e7a/stl/inc/execution#L249

And note that these values are for a single algorithm run, other calls to parallel algorithms will have its own chunks
Sure TBB will not use the internal `<execution>`  machinery, so it will use its own accounting.

`<execution>` uses Windows Thread Pool with all defaults, so it spawns up to 512 threads shared between all algorithms and other clients of the thread pool. I don't know if TBB uses its own thread pool or relies on Windows thread pool; in the former case it may add threads on top of this 512 maximum.


Apparently the resource control in parallel algorithms is still unresolved problem. See DevCom-1611069
